### ISSUE TITLE
AssertionError while looking up variables

### ISSUE THREAD

**Issue by joedsilva** [numba/numba#3204]
## Reporting a bug

- [ X] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Traceback (most recent call last):
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/errors.py", line 577, in new_error_context
    yield
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/lowering.py", line 254, in lower_block
    self.lower_inst(inst)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/objmode.py", line 65, in lower_inst
    value = self.lower_assign(inst)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/objmode.py", line 159, in lower_assign
    return self.lower_expr(value)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/objmode.py", line 368, in lower_expr
    val = self.loadvar(expr.value.name)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/objmode.py", line 520, in loadvar
    assert name in self._live_vars, name
AssertionError: res

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bug3.py", line 17, in <module>
    res = loop(m,v,2)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/dispatcher.py", line 368, in _compile_for_args
    raise e
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/dispatcher.py", line 325, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/dispatcher.py", line 653, in compile
    cres = self._compiler.compile(args, return_type)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/dispatcher.py", line 83, in compile
    pipeline_class=self.pipeline_class)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 873, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 367, in compile_extra
    return self._compile_bytecode()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 804, in _compile_bytecode
    return self._compile_core()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 791, in _compile_core
    res = pm.run(self.status)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 253, in run
    raise patched_exception
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 245, in run
    stage()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 438, in stage_objectmode_frontend
    cres = self.frontend_looplift()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 428, in frontend_looplift
    lifted=tuple(loops), lifted_from=None)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 887, in compile_ir
    lifted_from=lifted_from)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 375, in compile_ir
    return self._compile_ir()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 811, in _compile_ir
    return self._compile_core()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 791, in _compile_core
    res = pm.run(self.status)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 253, in run
    raise patched_exception
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 245, in run
    stage()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 652, in stage_objectmode_backend
    self._backend(lowerfn, objectmode=True)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 628, in _backend
    lowered = lowerfn()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 601, in backend_object_mode
    self.flags)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/compiler.py", line 1018, in py_lowering_stage
    lower.lower()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/lowering.py", line 173, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/lowering.py", line 214, in lower_normal_function
    entry_block_tail = self.lower_function_body()
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/lowering.py", line 239, in lower_function_body
    self.lower_block(block)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/lowering.py", line 254, in lower_block
    self.lower_inst(inst)
  File "/usr/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/errors.py", line 585, in new_error_context
    six.reraise(type(newerr), newerr, tb)
  File "/home/jdsilv2/.local/lib/python3.5/site-packages/numba/six.py", line 659, in reraise
    raise value
numba.errors.LoweringError: Failed at object (object mode frontend)
Failed at object (object mode backend)
res

File "bug3.py", line 12:
def loop(m, v, cnt):
    <source elided>
	
	return res
 ^
[1] During: lowering "$41.2 = cast(value=res)" at bug3.py (12)
-------------------------------------------------------------------------------
This should not have happened, a problem has occurred in Numba's internals.

Please report the error message and traceback, along with a minimal reproducer
at: https://github.com/numba/numba/issues/new

If more help is needed please feel free to speak to the Numba core developers
directly at: https://gitter.im/numba/numba

Thanks in advance for your help in improving Numba!

```python
import numpy as np
from numba import jit 

def mul(m ,v):
  return m * v 

@jit
def loop(m, v, cnt):
  for i in range(0, cnt):
    res = mul(m,v)
  
  return res 

m = 2.0 
v = 3.0 

res = loop(m,v,2)
print(res)
```
The issue goes away if we apply jit on mul function as well OR if we include the the logic of mul directly in the definition of loop OR if we include res=None at the begining of the definition of loop.

**Comment by stuartarchibald** [comment]
Thanks for the report. I think this is another instance of liveness analysis failing in [`object mode`](http://numba.pydata.org/numba-doc/dev/user/5minguide.html#what-is-nopython-mode). If `mul` is called without being `@jit` decorated, it triggers the `loop` function to be compiled in [`object mode`](http://numba.pydata.org/numba-doc/dev/user/5minguide.html#what-is-nopython-mode), which then hits a [known bug](https://github.com/numba/numba/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+liveness) in liveness analysis. Inlining `mul` in `loop` will not trigger it as [`nopython`](http://numba.pydata.org/numba-doc/dev/user/5minguide.html#what-is-nopython-mode) mode will be in use, same applies to a `@jit` decorated `mul`.

**Comment by seibert** [comment]
Closed by #3230.
